File: leetcode/chain/23.cpp
#include <iostream>
#include <vector>

#include "list_node.h"

class Solution {
 public:
  ListNode* merge_2_lists(ListNode* l1, ListNode *l2) {
    ListNode *ret = new ListNode();
    ListNode *last = ret;
    while (l1 != nullptr && l2 != nullptr) {
      if (l1->val <= l2->val) {
        last->next = l1;
        l1 = l1->next;
      } else {
        last->next = l2;
        l2 = l2->next;
      }
      last = last->next;
    }

    last->next = l1 ? l1 : l2;
    return ret->next;
  }

  ListNode* merge_k_lists(std::vector<ListNode*>& lists, int l, int r) {
    if (l >= r) {
      return lists[l];
    }
    int m = l + (r - l) / 2;
    ListNode *l1 = merge_k_lists(lists, 0, m);
    ListNode *l2 = merge_k_lists(lists, m + 1, r);
    return merge_2_lists(l1, l2);
  }

  ListNode* merge_k_lists(std::vector<ListNode*> lists) {
    return merge_k_lists(lists, 0, lists.size() - 1);
  }
};

int main() {
  ListNode* node = new ListNode(3);
  node->append(new ListNode(4));
  node->append(new ListNode(5));
  node->print();

  ListNode* node2 = new ListNode(1);
  node2->append(new ListNode(2));
  node2->append(new ListNode(4));
  node2->print();

  ListNode* node3 = new ListNode(5);
  node3->append(new ListNode(6));
  node3->append(new ListNode(7));
  node3->print();

  std::vector<ListNode *> lists;
  lists.emplace_back(node);
  lists.emplace_back(node2);
  lists.emplace_back(node3);

  Solution solution = Solution();
  ListNode* ret = solution.merge_k_lists(lists);
  ret->print();
}

File: leetcode/chain/23.py
# Definition for singly-linked list.
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def merge2Lists(self, l1: ListNode, l2: ListNode) -> ListNode:
        ret = ListNode()
        last = ret
        while l1 and l2:
            if l1.val <= l2.val:
                last.next = l1
                l1 = l1.next
            else:
                last.next = l2
                l2 = l2.next
            last = last.next
        last.next = l1 if l1 is not None else l2
        return ret.next

    def mergeKLists_all(self, lists: List[ListNode]) -> ListNode:
        '''合并K个有序链表'''
        if len(lists) == 0 or lists is None:
            return None
        for i in range(1, len(lists)):
            lists[i] = self.merge2Lists(lists[i - 1], lists[i])
        return lists[len(lists) - 1]

    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        '''合并K个有序链表
        
        @Note:
            归并排序
        '''
        if len(lists) == 0 or lists is None:
            return None
        return self.mergeKLists_sort(lists, 0, len(lists) - 1)

    def mergeKLists_sort(self, lists: List[ListNode], l, r) -> ListNode:
        if l == r:
            return lists[l]  #注意返回值,merge2Lists会更改原链表
        m = (l + r) // 2
        l1 = self.mergeKLists_sort(lists, l, m)
        l2 = self.mergeKLists_sort(lists, m + 1, r)
        return self.merge2Lists(l1, l2)
